Transcoder Claimmanager Potential Race Condition

**Describe the bug**
When you restart the claimmanager, with some unsubmitted claims, sometimes the node crashes with the following error:

```
8 7270 8087 8433 8614 8634 7823 8006 8724 8889 9227 8988 7297 7832 8148 8858 11211 6978 8257 10539 10617 8090 8445 7272 7929 8821 9619 10960 7801 8146 8572 9131 10785 11191 7381 10268 7933 7947 9560 10646 7915 8082 9327 9341 10156 11220 11279 8587 9073 7489 7938 8042 8147 8161 8429 9859 10876 10778 8071 8318 8458 8815 10396 10436 7588 10378 10574 10894 9417 10298 11221 8166 10257 10260 10286 10982 7623 7164 7282 9348 7285 7587 8777 9135 7121 7271 8801 9809 10171 11174 6720 8094 9031 9788 7867 8226 9346 9713 10238 10889 6949 7741 7748 10151 10205 10620 8766 8835 11293 8932 9252 8145 10640 6743 6955 8770 9480 8353 7102 10118 7119 7745 8279 10217 10452 7818 9076 7395 8046 8229 10330 7126 8661 7941 7969 8692 10909 11003 9412 9450 9558 9519 10675 9747 11252 11379 7946 8283 9524 10322 8128 8637 10486 10776 8628 8929 9710 8281 9692 10495 11057 11261 11270 7239 8346 8709 9138 10766 9225 10952 10903 7387 7913 8826 8832 9153 9765 6992 8771 9662 9778 7006 7470 8240 9391 10347 10647 8221 8233 8956 9306 11284 6758 7790 8301 10358 7509 9598 11250 7935 8358 9148 10588 10956 11192 6976 10382 10469 6993 7499 7974 8242 9295 11223 11365 7294 7300 7471 7975 10248 10356 6721 7405 8045 8776 9423 8440 9105 10145 10178 11271 9263 10530 7487 8828 8939 10761 11291 7815 7878 8209 8942 9460 8307 10249 8487 9544 9620 10473 8262 9603 8491 11356 6778 7389 8213 8486 9233 10505 10361 11312 7198 7523 8743 9686 10472 9737 7221 7536 8442 8906 9074 9857 8726 9292 6752 7724 7854 9059 10360 7203 7260 9116 9175 9421 7612 6768 7328 9235 7205 10639 7036 8204 9506 10126 11203 11413 7932 8239 10763 8266 8463 9494 10144 10414 10623 7206 9160 9368 9312 9640 10214 7907 9789 9895 7263 7831 9668 10437 10482 7278 8234 9835 6943 7579 7784 8108 8122 9673 10954 11059 7256 8093 8453 8800 9128 10262 7859 8461 8946 9564 7800 9443 9447 9764 9042 9648 10320 7403 8218 8187 8182 8767 10631 11259 6947 8100 8323 8983 10239 9688 11265 7235 7993 8959 9651 11366 11406 10439 9379 9468 10567 10582 10974 7117 7980 8326 8454 8649 9319 7251 7845 7952 8151 9172 10373 10428 10523 7770 8424 8849 6770 7546 8152 8517 9376 9518 8903 10461 8119 9180 9671 10665 11320 9075 9096 9336 9467 10658 11397 8241 9454 9816 7504 8113 9210 9870 10670 8386 8387 8549 10308 10891 7224 8004 8014 8456 9476 9734 8154 8275 10196 11363 9066 7303 8732 8896 9271 9695 8013 9805 10159 11167 9582 10413 7808 8031 8292 8865 8909 9097 9807 9846 10755 10932 8072 8802 9240 9245 10269 7492 7799 8586 10985 9774 10981 11349 9169 10594 7625 8427 9089 9629 10667 7398 8773 7433 8219 9900 10540 11357 6905 8298 8968 8987 8136 8364 8468 10368 6946 7281 7372 8271 9822 11316 9178 9893 11371 7393 8989 9744] ranges: [[6704 6705] [6715 6715] [6717 6736] [6738 6741] [6743 6786] [6882 6895] [6897 6906] [6930 6931] [6941 6961] [6963 7021] [7023 7037] [7090 7091] [7094 7107] [7109 7113] [7115 7126] [7164 7164] [7198 7198] [7203 7218] [7220 7224] [7226 7335] [7354 7408] [7433 7435] [7443 7443] [7467 7546] [7578 7580] [7582 7584] [7586 7621] [7623 7625] [7690 7690] [7692 7692] [7721 7725] [7727 7731] [7733 7735] [7737 7808] [7812 7813] [7815 7816] [7818 7836] [7838 7841] [7843 7847] [7849 7857] [7859 7863] [7865 7871] [7878 7888] [7890 7982] [7984 9903] [10005 10006] [10114 10114] [10116 10560] [10564 10677] [10751 10752] [10754 10792] [10814 10814] [10862 11010] [11051 11052] [11054 11060] [11124 11127] [11164 11379] [11396 11400] [11405 11406] [11408 11420] [11424 11424]]
I0726 20:52:43.191990   19754 db.go:543] Inserting claim for job 2768
I0726 20:52:43.191326   19754 cert.go:83] Private key and cert not found. Generating
I0726 20:52:43.192356   19754 cert.go:22] Generating cert for 13.58.143.99
I0726 20:52:43.192411   19754 db.go:560] Guessed claim ID to be 25 for job 2768
I0726 20:52:43.192645   19754 rpc.go:402] Listening for RPC on :4433
I0726 20:52:43.188703   19754 db.go:560] Guessed claim ID to be 4 for job 2797
E0726 20:52:43.192978   19754 db.go:563] Unable to insert claim database is locked
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd060f8]

goroutine 502 [running]:
github.com/livepeer/go-livepeer/eth.(*BasicClaimManager).ClaimVerifyAndDistributeFees(0xc4200eedc0, 0xc4209577a0, 0x2)
        /home/ubuntu/go/src/github.com/livepeer/go-livepeer/eth/claimmanager.go:339 +0x1238
github.com/livepeer/go-livepeer/eth.RecoverClaims.func1(0xc430736b00, 0xc4200eedc0)
        /home/ubuntu/go/src/github.com/livepeer/go-livepeer/eth/claimmanager.go:141 +0xd3
created by github.com/livepeer/go-livepeer/eth.RecoverClaims
        /home/ubuntu/go/src/github.com/livepeer/go-livepeer/eth/claimmanager.go:139 +0x592
```

**To Reproduce**
Steps to reproduce the behavior:
1. Start a stream, make sure a transcoder starts doing the work.
2. Stop the stream, restart the transcoder

**Expected behavior**
Node not to crash.  This doesn't happen all the time, which leads me to suspect it's a race condition.

--- a/cmd/livepeer_cli/wizard_bond.go
+++ b/cmd/livepeer_cli/wizard_bond.go
@@ -170,6 +170,9 @@ func (w *wizard) bond() {
 	for amount.Cmp(big.NewInt(0)) == 0 || balBigInt.Cmp(amount) < 0 {
 		fmt.Printf("Enter bond amount - ")
 		amount = w.readBigInt()
+		if amount.Cmp(big.NewInt(0)) == 0 {
+			break
+		}
 		if balBigInt.Cmp(amount) < 0 {
 			fmt.Printf("Must enter an amount smaller than the current balance. ")
 		}

--- a/cmd/livepeer_cli/wizard_stats.go
+++ b/cmd/livepeer_cli/wizard_stats.go
@@ -29,7 +29,6 @@ func (w *wizard) stats(showTranscoder bool) {
 
 	table := tablewriter.NewWriter(os.Stdout)
 	data := [][]string{
-		[]string{"Node Addr", w.getNodeAddr()},
 		[]string{"HTTP Port", w.httpPort},
 		[]string{"Controller Address", addrMap["Controller"].Hex()},
 		[]string{"LivepeerToken Address", addrMap["LivepeerToken"].Hex()},
@@ -257,10 +256,6 @@ func (w *wizard) delegatorStats() {
 	table.SetRowLine(true)
 	table.SetColumnSeparator("|")
 	table.Render()
-}
-
-func (w *wizard) getNodeAddr() string {
-	return httpGet(fmt.Sprintf("http://%v:%v/nodeAddrs", w.host, w.httpPort))
 }
 
 func (w *wizard) getProtocolParameters() (lpTypes.ProtocolParameters, error) {

--- a/common/db.go
+++ b/common/db.go
@@ -189,6 +189,10 @@ func InitDB(dbPath string) (*DB, error) {
 		glog.Error("Unable to open DB ", dbPath, err)
 		return nil, err
 	}
+	// The DB connection might be used in multiple goroutines (i.e. when recovering claims during node restart)
+	// resulting in concurrent access. SQLite can only handle one writer at a time, so if concurrent writes occur
+	// we can encounter a `database is locked` error. To avoid concurrent writes, we limit SQLite to a single connection
+	db.SetMaxOpenConns(1)
 	d.dbh = db
 	schemaBuf := new(bytes.Buffer)
 	tmpl := template.Must(template.New("schema").Parse(schema))

--- a/eth/claimmanager.go
+++ b/eth/claimmanager.go
@@ -335,7 +335,11 @@ func (c *BasicClaimManager) ClaimVerifyAndDistributeFees() error {
 		claimID := c.claims
 		if c.db != nil {
 			// check db in case we had any concurrent claims
-			claimIDp, _ := c.db.InsertClaim(c.jobID, segRange, root.Hash)
+			claimIDp, err := c.db.InsertClaim(c.jobID, segRange, root.Hash)
+			if err != nil {
+				glog.Errorf("Job %v Error: %v - inserting claim into DB", c.jobID, err)
+				continue
+			}
 			claimID = *claimIDp
 		}
 

--- a/server/mediaserver.go
+++ b/server/mediaserver.go
@@ -4,6 +4,7 @@ Package server is the place we integrate the Livepeer node with the LPMS media s
 package server
 
 import (
+	"bytes"
 	"context"
 	"crypto/tls"
 	"errors"
@@ -216,10 +217,26 @@ func gotRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 				glog.Error("Unable to find active broadcasts ", err)
 			}
 
+			getReusableBroadcast := func(b *common.DBJob) (*ethTypes.Job, bool) {
+				job := common.DBJobToEthJob(b)
+
+				// Transcoder must still be registered
+				if t, err := s.LivepeerNode.Eth.GetTranscoder(b.Transcoder); err != nil || t.Status == "Not Registered" {
+					return job, false
+				}
+
+				// Transcode options must be the same
+				if bytes.Compare(common.ProfilesToTranscodeOpts(job.Profiles), common.ProfilesToTranscodeOpts(BroadcastJobVideoProfiles)) != 0 {
+					return job, false
+				}
+
+				// Existing job with a max price per segment <= current desired price is acceptable
+				return job, job.MaxPricePerSegment.Cmp(BroadcastPrice) <= 0
+			}
+
 			for _, b := range bcasts {
-				// check if assigned transcoder is still valid.
-				if _, err := s.LivepeerNode.Eth.GetTranscoder(b.Transcoder); err == nil {
-					job := common.DBJobToEthJob(b)
+				// check if job has acceptable price, transcoding options and valid assigned transcoder
+				if job, reusable := getReusableBroadcast(b); reusable {
 					rpcBcast, err = s.startBroadcast(job, manifest, nonce)
 					if err == nil {
 						startSeq = int(b.Segments) + 1

